hw/acpi/aml-build.c

```diff
@@ -2016,22 +2027,25 @@ acpi_build_mcfg(GArray *table_data, BIOSLinker *linker, AcpiMcfgInfo *info)
 {
     AcpiTableMcfg *mcfg;
     const char *sig;
-    int len = sizeof(*mcfg) + 1 * sizeof(mcfg->allocation[0]);
+    int len = sizeof(*mcfg) + info->segment_total * sizeof(mcfg->allocation[0]);
+    uint16_t i;
 
     mcfg = acpi_data_push(table_data, len);
-    mcfg->allocation[0].address = cpu_to_le64(info->mcfg_base);
-    /* Only a single allocation so no need to play with segments */
-    mcfg->allocation[0].pci_segment = cpu_to_le16(0);
-    mcfg->allocation[0].start_bus_number = 0;
-    mcfg->allocation[0].end_bus_number = PCIE_MMCFG_BUS(info->mcfg_size - 1);
+    for (i = 0; i < info->segment_total; i++) {
+        mcfg->allocation[i].address = cpu_to_le64(info->mcfg_base[i]);
+        mcfg->allocation[i].pci_segment = cpu_to_le16(i);
+        mcfg->allocation[i].start_bus_number = 0;
+        mcfg->allocation[i].end_bus_number = PCIE_MMCFG_BUS(info->mcfg_size[i] - 1);
+    }
 
     /* MCFG is used for ECAM which can be enabled or disabled by guest.
      * To avoid table size changes (which create migration issues),
      * always create the table even if there are no allocations,
      * but set the signature to a reserved value in this case.
      * ACPI spec requires OSPMs to ignore such tables.
      */
-    if (info->mcfg_base == PCIE_BASE_ADDR_UNMAPPED) {
+    /* TODO: What if some of pci_host has the UNMAPPED addr? */
```

Comment:
@liujing2 this is a non issue here. This check addresses a different condition.

Reply:
@mcastelino  I am not sure  if mcfg_base[0] or mcfg_base[1] == UNMAPPED, I mean if we need check all?